### PR TITLE
[CINFRA-136] Replicated state machine

### PR DESCRIPTION
**Pull request by maierlars**
### Scope & Purpose

Add replicated state machine on top of replicated logs. It is designed to make implementing state machines simple and reliable.

### CODE REVIEW

**Comment by apetenchea** [comment]
I noticed the build was failing and added a commit to your PR.

File: tests/Replication2/ReplicatedState/ReplicatedStateTest.cpp [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+      feature->createReplicatedState("my-state", log));
+  ASSERT_NE(state, nullptr);
+
+  // create a leader in term 1
+  auto leaderLog = makeReplicatedLog(LogId{1});
+  auto leader = leaderLog->becomeLeader(LogConfig(2, 2, 2, false), "leader",
+                                        LogTerm{1}, {follower});
+  auto mux = streams::LogMultiplexer<ReplicatedStateStreamSpec<MyState>>::construct(leader);
+  auto inputStream = mux->getStreamById<1>();
+  inputStream->insert({.key = "hello", .value = "world"});
```
Comment (apetenchea) [inline tests/Replication2/ReplicatedState/ReplicatedStateTest.cpp:90]: This looks like it wouldn't affect the test in any way. Just pointing it out, not sure if it's left intentionally.

**Review by apetenchea — APPROVED** [review]
Looked over the code and double checked the tests. Apart from my remark regarding LogMultiplexer.tpp#297, the rest are style guide considerations and "make-sure's". Nothing major though, could be merged.

File: arangod/Replication2/ReplicatedLog/NetworkMessages.h [hunk 2]
```diff
@@ -51,11 +51,11 @@ class Result;
 
 namespace arangodb::replication2::replicated_log {
 
-struct MessageId : implement_compare<MessageId> {
+struct MessageId {
```
[the remark this answers is not in the document]
Comment (maierlars) [reply arangod/Replication2/ReplicatedLog/NetworkMessages.h]: Already merged in devel.

File: arangod/Replication2/ReplicatedState/ReplicatedState.tpp [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+    demux->listen();
+    stream = demux->template getStreamById<1>();
+
+    LOG_TOPIC("1d843", TRACE, Logger::REPLICATED_STATE)
+        << "creating follower state instance";
+    state = locked->factory->constructFollower();
+
+    LOG_TOPIC("ea777", TRACE, Logger::REPLICATED_STATE)
+        << "check if new snapshot is required";
+
```
[the remark this answers is not in the document]
Comment (maierlars) [reply arangod/Replication2/ReplicatedState/ReplicatedState.tpp:292]: Yes, but I postponed it into another [ticket](https://arangodb.atlassian.net/browse/CINFRA-161). This PR is already big enough.

File: arangod/Replication2/ReplicatedState/StateCommon.h [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+  [[nodiscard]] auto saturatedDecrement(uint64_t delta = 1) const noexcept -> StateGeneration;
+
+  friend auto operator<=>(StateGeneration const&, StateGeneration const&) = default;
+
+  [[nodiscard]] auto operator+(std::uint64_t delta) const -> StateGeneration;
+
+  friend auto operator<<(std::ostream&, StateGeneration) -> std::ostream&;
+
+  [[nodiscard]] explicit operator velocypack::Value() const noexcept;
```
[the remark this answers is not in the document]
Comment (maierlars) [reply arangod/Replication2/ReplicatedState/StateCommon.h:52]: Yes, when needed.

File: arangod/Replication2/ReplicatedState/StateStatus.cpp [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+  TRI_ASSERT(false) << "invalid state value " << int(state);
+  return "(unknown-internal-follower-state)";
+}
+
+void StateStatus::toVelocyPack(velocypack::Builder& builder) const {
+  std::visit([&](auto const& s) { s.toVelocyPack(builder); }, variant);
+}
+
+auto StateStatus::fromVelocyPack(velocypack::Slice slice) -> StateStatus {
+  auto role = slice.get(String_Role).stringView();
```
[the remark this answers is not in the document]
Comment (maierlars) [reply arangod/Replication2/ReplicatedState/StateStatus.cpp:136]: This applies to all velocypack deserializers we ever wrote in the past. That's why we have all those `Expected Uint` exceptions. And that's the reason why I really would like to have a serializer/deserializer framework.

File: tests/Replication2/ReplicatedState/ReplicatedStateTest.cpp (hunk 1 above)
Comment (maierlars) [reply tests/Replication2/ReplicatedState/ReplicatedStateTest.cpp:90]: oh it does. We assert that this entry was applied to the state in line 110 to 111.

**Pull request merged**